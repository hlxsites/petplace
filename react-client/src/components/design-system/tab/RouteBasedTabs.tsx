import type { FC } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import type { Tab, TabsProps } from "./Tabs";
import { Tabs } from "./Tabs";

export type RouteTab = Tab & {
  exactRoute?: boolean;
  route: string;
};

type RouteBasedTabsProps = Omit<TabsProps, "onChange" | "tabs" | "value"> & {
  tabs: RouteTab[];
};

export const RouteBasedTabs: FC<RouteBasedTabsProps> = ({ tabs }) => {
  const navigate = useNavigate();
  const { pathname } = useLocation();

  if (tabs.length < 1) {
    return null;
  }

  const selectedTab =
    tabs.find(({ exactRoute, route }) =>
      exactRoute ? route === pathname : pathname.includes(route)
    )?.label || tabs[0].label;

  return (
    <Tabs onChange={onChangeSelectedTab} tabs={tabs} value={selectedTab} />
  );

  function onChangeSelectedTab(newSelectedTab: string) {
    const tab = tabs.find(({ label }) => label === newSelectedTab);
    if (tab) navigate(tab.route);
  }
};
