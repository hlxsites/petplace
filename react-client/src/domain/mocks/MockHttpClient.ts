import {
  HttpClientRepository,
  HttpResponse,
} from "../repository/HttpClientRepository";

export class MockHttpClient implements HttpClientRepository {
  private data: unknown;
  private error: unknown;

  constructor({ data, error }: { data?: unknown; error?: unknown } = {}) {
    this.data = data;
    this.error = error;
  }

  delete(): Promise<HttpResponse> {
    if (this.error) {
      // @ts-expect-error - this is a mock implementation
      return { error: this.error };
    }

    // @ts-expect-error - this is a mock implementation
    return { data: this.data };
  }

  get(): Promise<HttpResponse> {
    if (this.error) {
      // @ts-expect-error - this is a mock implementation
      return { error: this.error };
    }

    // @ts-expect-error - this is a mock implementation
    return { data: this.data };
  }

  post(): Promise<HttpResponse> {
    if (this.error) {
      // @ts-expect-error - this is a mock implementation
      return { error: this.error };
    }

    // @ts-expect-error - this is a mock implementation
    return { data: this.data };
  }

  postFormData(): Promise<HttpResponse> {
    if (this.error) {
      // @ts-expect-error - this is a mock implementation
      return { error: this.error };
    }

    // @ts-expect-error - this is a mock implementation
    return { data: this.data };
  }
}
