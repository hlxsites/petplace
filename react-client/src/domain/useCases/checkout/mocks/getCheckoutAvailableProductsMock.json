{
  "membershipProducts": {
    "annualMembership": {
      "itemName": "Test item name",
      "itemId": "Test item id",
      "itemType": "Test item type",
      "salesPrice": "29.99",
      "renewPrice": "9.99",
      "additionalProductList": [
        {
          "itemName": "Test additional product name",
          "itemId": "Test additional product id",
          "itemType": "Test additional product item type",
          "price": "39.99",
          "uiName": "Test ui name"
        }
      ]
    },

    "subscriptions": null,
    "tags": null
    
  }
}