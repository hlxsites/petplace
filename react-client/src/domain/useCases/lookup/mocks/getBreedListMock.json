[
  { "id": 1, "name": "Poodle" },
  { "id": 2, "name": "Golden Retriever" },
  { "id": 3, "name": "Labrador" },
  { "id": 4, "name": "Pug" },
  { "id": 5, "name": "Beagle" }
]
