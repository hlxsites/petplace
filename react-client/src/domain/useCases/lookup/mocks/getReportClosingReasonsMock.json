[
  {
    "id": 4,
    "name": "I picked up my pet from the shelter/finder"
  },
  {
    "id": 2,
    "name": "I found my pet myself"
  },
  {
    "id": 11,
    "name": "I surrendered my pet"
  },
  {
    "id": 1,
    "name": "My pet was never lost"
  },
  {
    "id": 18,
    "name": "I do not own this pet"
  },
  {
    "id": 3,
    "name": "This pet is deceased"
  }
]