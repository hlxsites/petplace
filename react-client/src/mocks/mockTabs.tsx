export function getTestTabs(id: string = "123") {
  return [
    {
      content: () => <div data-testid="CharacteristicsTabContent" />,
      exactRoute: true,
      label: "Characteristics",
      route: "/characteristics/" + id,
    },
    {
      content: () => <div data-testid="NotesTabContent" />,
      label: "Notes",
      route: "/notes/" + id,
    },
    {
      content: () => <div data-testid="TimelineTabContent" />,
      exactRoute: true,
      label: "Timeline",
      route: "/timeline/" + id,
    },
  ];
}
