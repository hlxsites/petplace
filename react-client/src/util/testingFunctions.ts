import { screen } from "@testing-library/react";

export function getSelectedTab() {
  return screen.getByRole("tab", { selected: true }).textContent;
}

export async function findByTextContent(
  textMatch: string | RegExp
): Promise<HTMLElement> {
  // @ts-expect-error - TS doesn't understand what we want with the utility here
  return screen.findByText(findByContext(textMatch));
}

export function getByTextContent(textMatch: string | RegExp): HTMLElement {
  // @ts-expect-error - TS doesn't understand what we want with the utility here
  return screen.getByText(findByContext(textMatch));
}

export type MatcherFunction = (
  content: string,
  element: Element | null
) => boolean;

function findByContext(textMatch: string | RegExp) {
  return (_: string, node: Element) => {
    const hasText = (inNode: Element): boolean => {
      if (typeof textMatch === "string") {
        return inNode.textContent === textMatch;
      }
      return inNode.textContent?.match(textMatch) !== null;
    };

    const nodeHasText = hasText(node);
    const childrenDontHaveText = Array.from(node?.children || []).every(
      (child) => !hasText(child)
    );
    return nodeHasText && childrenDontHaveText;
  };
}
