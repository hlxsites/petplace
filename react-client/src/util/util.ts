import clsx, { type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function classNames(...inputs: ClassValue[]): string {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call
  return twMerge(clsx(inputs));
}
