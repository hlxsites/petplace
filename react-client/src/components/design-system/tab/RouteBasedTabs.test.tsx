import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import type { ComponentProps } from "react";
import { MemoryRouter } from "react-router-dom";
import { getTestTabs } from "~/mocks/mockTabs";
import { getSelectedTab } from "~/util/testingFunctions";
import { RouteBasedTabs } from "./RouteBasedTabs";

const { getAllByRole, getByTestId, getByText, queryByRole } = screen;

describe("RouteBasedTabs", () => {
  it("should display nothing", () => {
    getRenderer({
      tabs: [],
    });
    expect(queryByRole("tab")).not.toBeInTheDocument();
  });

  it("should display tabs", () => {
    const tabs = getTestTabs("123");
    getRenderer({
      tabs,
    });
    tabs.forEach(({ label }) => expect(getByText(label)).toBeInTheDocument());
    expect(getAllByRole("tab")).toHaveLength(tabs.length);
  });

  it.each([
    ["Characteristics", "/characteristics/123"],
    ["Notes", "/notes/123"],
    ["Timeline", "/timeline/123"],
  ])(
    "should start with %s Tab selected when route='%s'",
    (expected, initialRoute) => {
      const tabs = getTestTabs("123");
      getRenderer({
        tabs,
        initialRoute,
      });

      expect(getSelectedTab()).toBe(expected);
    }
  );

  it("should allow user to select tabs", async () => {
    const tabs = getTestTabs();
    getRenderer({
      tabs,
    });

    expect(tabs).toHaveLength(3);

    await userEvent.click(getByText(tabs[2].label));
    expect(getSelectedTab()).toBe(tabs[2].label);

    await userEvent.click(getByText(tabs[1].label));
    expect(getSelectedTab()).toBe(tabs[1].label);
  });

  it("should render first tab content", () => {
    const tabs = getTestTabs();
    getRenderer({
      tabs,
      initialRoute: tabs[0].route,
    });
    expect(getByTestId("CharacteristicsTabContent")).toBeInTheDocument();
  });

  it("should render second tab content", () => {
    const tabs = getTestTabs("222");
    getRenderer({
      tabs,
      initialRoute: tabs[1].route,
    });
    expect(getByTestId("NotesTabContent")).toBeInTheDocument();
  });
});

function getRenderer({
  initialRoute,
  ...props
}: ComponentProps<typeof RouteBasedTabs> & { initialRoute?: string }) {
  return render(
    <MemoryRouter initialEntries={[initialRoute || "/characteristics/123"]}>
      <RouteBasedTabs {...props} />
    </MemoryRouter>
  );
}
