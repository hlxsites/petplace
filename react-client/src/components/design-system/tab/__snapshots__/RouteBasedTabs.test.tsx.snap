// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RouteBasedTabs /> should allow user to select tabs 1`] = `
<div>
  <div
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="flex"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-:rg:-content-Characteristics"
        aria-label="Tab option: Characteristics"
        aria-selected="false"
        class="flex w-full justify-center rounded-none border-0 border-b-[3px] border-b-neutral-300 bg-transparent p-small text-base font-bold text-neutral-500 outline-none hover:border-0 hover:border-b-[3px] hover:border-b-neutral-300 hover:bg-transparent hover:outline-none focus:bg-transparent focus:outline-none aria-selected:border-b-purple-500 aria-selected:text-purple-500"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:rg:-trigger-Characteristics"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex place-items-center gap-small"
        >
          <p
            class="inline-block text-black font-franklin font-normal text-base leading-6"
          >
            Characteristics
          </p>
        </div>
      </button>
      <button
        aria-controls="radix-:rg:-content-Notes"
        aria-label="Tab option: Notes"
        aria-selected="false"
        class="flex w-full justify-center rounded-none border-0 border-b-[3px] border-b-neutral-300 bg-transparent p-small text-base font-bold text-neutral-500 outline-none hover:border-0 hover:border-b-[3px] hover:border-b-neutral-300 hover:bg-transparent hover:outline-none focus:bg-transparent focus:outline-none aria-selected:border-b-purple-500 aria-selected:text-purple-500"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:rg:-trigger-Notes"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex place-items-center gap-small"
        >
          <p
            class="inline-block text-black font-franklin font-normal text-base leading-6"
          >
            Notes
          </p>
        </div>
      </button>
      <button
        aria-controls="radix-:rg:-content-Timeline"
        aria-label="Tab option: Timeline"
        aria-selected="true"
        class="flex w-full justify-center rounded-none border-0 border-b-[3px] border-b-neutral-300 bg-transparent p-small text-base font-bold text-neutral-500 outline-none hover:border-0 hover:border-b-[3px] hover:border-b-neutral-300 hover:bg-transparent hover:outline-none focus:bg-transparent focus:outline-none aria-selected:border-b-purple-500 aria-selected:text-purple-500"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-:rg:-trigger-Timeline"
        role="tab"
        tabindex="0"
        type="button"
      >
        <div
          class="flex place-items-center gap-small"
        >
          <p
            class="inline-block text-black font-franklin font-normal text-base leading-6"
          >
            Timeline
          </p>
        </div>
      </button>
    </div>
    <div
      aria-label="Tab content of: Characteristics"
      aria-labelledby="radix-:rg:-trigger-Characteristics"
      class="pt-large"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:rg:-content-Characteristics"
      role="tabpanel"
      style="animation-duration: 0s;"
      tabindex="0"
    />
    <div
      aria-label="Tab content of: Notes"
      aria-labelledby="radix-:rg:-trigger-Notes"
      class="pt-large"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:rg:-content-Notes"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-label="Tab content of: Timeline"
      aria-labelledby="radix-:rg:-trigger-Timeline"
      class="pt-large"
      data-orientation="horizontal"
      data-state="active"
      id="radix-:rg:-content-Timeline"
      role="tabpanel"
      tabindex="0"
    >
      <div
        data-testid="TimelineTabContent"
      />
    </div>
  </div>
</div>
`;

exports[`<RouteBasedTabs /> should allow user to select tabs 2`] = `
<div>
  <div
    data-orientation="horizontal"
    dir="ltr"
  >
    <div
      aria-orientation="horizontal"
      class="flex"
      data-orientation="horizontal"
      role="tablist"
      style="outline: none;"
      tabindex="0"
    >
      <button
        aria-controls="radix-:rg:-content-Characteristics"
        aria-label="Tab option: Characteristics"
        aria-selected="false"
        class="flex w-full justify-center rounded-none border-0 border-b-[3px] border-b-neutral-300 bg-transparent p-small text-base font-bold text-neutral-500 outline-none hover:border-0 hover:border-b-[3px] hover:border-b-neutral-300 hover:bg-transparent hover:outline-none focus:bg-transparent focus:outline-none aria-selected:border-b-purple-500 aria-selected:text-purple-500"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:rg:-trigger-Characteristics"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex place-items-center gap-small"
        >
          <p
            class="inline-block text-black font-franklin font-normal text-base leading-6"
          >
            Characteristics
          </p>
        </div>
      </button>
      <button
        aria-controls="radix-:rg:-content-Notes"
        aria-label="Tab option: Notes"
        aria-selected="true"
        class="flex w-full justify-center rounded-none border-0 border-b-[3px] border-b-neutral-300 bg-transparent p-small text-base font-bold text-neutral-500 outline-none hover:border-0 hover:border-b-[3px] hover:border-b-neutral-300 hover:bg-transparent hover:outline-none focus:bg-transparent focus:outline-none aria-selected:border-b-purple-500 aria-selected:text-purple-500"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="active"
        id="radix-:rg:-trigger-Notes"
        role="tab"
        tabindex="0"
        type="button"
      >
        <div
          class="flex place-items-center gap-small"
        >
          <p
            class="inline-block text-black font-franklin font-normal text-base leading-6"
          >
            Notes
          </p>
        </div>
      </button>
      <button
        aria-controls="radix-:rg:-content-Timeline"
        aria-label="Tab option: Timeline"
        aria-selected="false"
        class="flex w-full justify-center rounded-none border-0 border-b-[3px] border-b-neutral-300 bg-transparent p-small text-base font-bold text-neutral-500 outline-none hover:border-0 hover:border-b-[3px] hover:border-b-neutral-300 hover:bg-transparent hover:outline-none focus:bg-transparent focus:outline-none aria-selected:border-b-purple-500 aria-selected:text-purple-500"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="inactive"
        id="radix-:rg:-trigger-Timeline"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <div
          class="flex place-items-center gap-small"
        >
          <p
            class="inline-block text-black font-franklin font-normal text-base leading-6"
          >
            Timeline
          </p>
        </div>
      </button>
    </div>
    <div
      aria-label="Tab content of: Characteristics"
      aria-labelledby="radix-:rg:-trigger-Characteristics"
      class="pt-large"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:rg:-content-Characteristics"
      role="tabpanel"
      style=""
      tabindex="0"
    />
    <div
      aria-label="Tab content of: Notes"
      aria-labelledby="radix-:rg:-trigger-Notes"
      class="pt-large"
      data-orientation="horizontal"
      data-state="active"
      id="radix-:rg:-content-Notes"
      role="tabpanel"
      tabindex="0"
    >
      <div
        data-testid="NotesTabContent"
      />
    </div>
    <div
      aria-label="Tab content of: Timeline"
      aria-labelledby="radix-:rg:-trigger-Timeline"
      class="pt-large"
      data-orientation="horizontal"
      data-state="inactive"
      hidden=""
      id="radix-:rg:-content-Timeline"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
